### ISSUE TITLE
Function definitions should be recognized in terms of any compound command

### ISSUE THREAD

**Issue by jcbollinger** [koalaman/shellcheck#1319]
POSIX specifies that shell functions can be defined in terms of *any* compound command.  By far the most common used to define shell functions is a brace-enclosed list, but parentheses-enclosed lists, `for` and `while` loops, `if` statements, and `case` statements are all compound commands, too.  Shellcheck accepts the brace-enclosed and parentheses-enclosed list variations, but rejects the others -- it cannot even parse them.  At least some real-world shells do accept them, per POSIX specifications.  Bash, for example, does so.

- Rule SC1064
- online
- [X] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

test_func() case $1 in
  x*) echo "starts with x" ;;
  *)  echo "does not start with x" ;;
esac

test_func xray
test_func exit
```

#### Here's what shellcheck currently says:

Line 3:
test_func() case $1 in
^-- SC1073: Couldn't parse this function. Fix to allow more checks.
            ^-- SC1064: Expected a { to open the function definition.
            ^-- SC1072: Fix any mentioned problems and try again.

#### Here's what I wanted or expected to see:

No issues detected!

Alternatively, shellcheck might emit a warning, but it should certainly be able to parse such function definitions, since they are, in fact, valid.

**Comment by koalaman** [comment]
You're right, of course. It's tragic because the fix is easier than the current behavior, but it causes a significant empirical reduction in helpfulness. See #748 for more.
